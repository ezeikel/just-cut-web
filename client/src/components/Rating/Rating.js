import React, { Component } from 'react';
import styled from 'styled-components';

import { spinKeyframe } from '../../globalStyles';
import starIcon from '../../assets/icons/star.svg';
import goldStarIcon from '../../assets/icons/gold-star.svg';

const RatingWrapper = styled.form`
  display: grid;
  grid-template-columns: repeat(5, 35px) auto;
  align-items: center;
  &.submitted {
    //TODO: Maybe swap DOM with feedback message instead of this
    display: none;
  }
`;

const RatingInput = styled.input`
  display: none;
`;

const RatingLabel = styled.label`
  background-image: url(${starIcon});
  background-repeat: no-repeat;
  background-size: contain;
  background-position: center;
  text-indent: -9999px;
  height: 25px;
  &.active {
    background-image: url(${goldStarIcon});
    animation: ${spinKeyframe} 0.3s ease-in-out;
  }
`;

const RatingTotal = styled.span`
  padding-left: var(--spacing-medium);
`;

const SubmitRating = styled.button`
  display: none;
  &.active {
    display: block;
  }
`;

class Rating extends Component {
  state = {
    rating: 0
  }

  componentWillMount() {
    if (this.props.ratings && this.props.ratings.length > 0) {
      this.calculateAverageRating(this.props.ratings);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.ratings && nextProps.ratings.length > 0) {
      this.calculateAverageRating(nextProps.ratings);
    }
  }

  calculateAverageRating = (ratings) => {
    const sum = ratings.reduce((a, b) => a + b);
    const avg = sum / ratings.length;
    this.setState({
      rating: avg
    });
  }

  totalRatings = () => {
    if (this.props.ratings && this.props.ratings.length) {
      return `${this.props.ratings.length} Rating${this.props.ratings.length > 1 ? 's' : ''}`;
    }

    return `No ratings yet.`;
  }

  renderRating = () => {
    if (this.props.readonly) {
      return (
        <RatingWrapper>
          <RatingInput id="1" type="radio" value="1" />
          <RatingLabel className={this.state.rating >= 1 ? 'active' : ''} htmlFor="1">Abysmal</RatingLabel>
          <RatingInput id="2" type="radio" value="2" />
          <RatingLabel className={this.state.rating >= 2 ? 'active' : ''} htmlFor="2">Poor</RatingLabel>
          <RatingInput id="3" type="radio" value="3" />
          <RatingLabel className={this.state.rating >= 3 ? 'active' : ''} htmlFor="3">Alright</RatingLabel>
          <RatingInput id="4" type="radio" value="4" />
          <RatingLabel className={this.state.rating >= 4 ? 'active' : ''} htmlFor="4">Good</RatingLabel>
          <RatingInput id="5" type="radio" value="5" />
          <RatingLabel className={this.state.rating >= 5 ? 'active' : ''} htmlFor="5">Great</RatingLabel>
          <RatingTotal>
            {this.totalRatings()}
          </RatingTotal>
        </RatingWrapper>
      );
    }

    return (
      <RatingWrapper onSubmit={this.props.handleSubmit} className={this.props.submitted ? 'submitted' : ''}>
        <RatingInput id="1" type="radio" value="1" onClick={this.props.updateRating} />
        <RatingLabel className={this.props.rating >= 1 ? 'active' : ''} htmlFor="1">Abysmal</RatingLabel>
        <RatingInput id="2" type="radio" value="2" onClick={this.props.updateRating} />
        <RatingLabel className={this.props.rating >= 2 ? 'active' : ''} htmlFor="2">Poor</RatingLabel>
        <RatingInput id="3" type="radio" value="3" onClick={this.props.updateRating} />
        <RatingLabel className={this.props.rating >= 3 ? 'active' : ''} htmlFor="3">Alright</RatingLabel>
        <RatingInput id="4" type="radio" value="4" onClick={this.props.updateRating} />
        <RatingLabel className={this.props.rating >= 4 ? 'active' : ''} htmlFor="4">Good</RatingLabel>
        <RatingInput id="5" type="radio" value="5" onClick={this.props.updateRating} />
        <RatingLabel className={this.props.rating >= 5 ? 'active' : ''} htmlFor="5">Great</RatingLabel>
        <SubmitRating className={this.props.changed ? 'active' : ''}>Submit Rating</SubmitRating>
      </RatingWrapper>
    );
  }

  render() {
    return this.renderRating();
  }
}

export default Rating;
