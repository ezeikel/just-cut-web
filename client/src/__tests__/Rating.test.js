import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Rating from '../components/Rating/Rating';

describe('Rating Component', () => {
  it('renders without crashing', () => {
    shallow(<Rating />);
  });

  it('should match snapshot', () => {
    expect(shallowToJson(shallow(<Rating />))).toMatchSnapshot();
  });

  it('should render a <RatingWrapper/>', () => {
    expect(shallow(<Rating />).find('RatingWrapper').length).toEqual(1);
  });

  it('should render five <RatingInput/>', () => {
    expect(shallow(<Rating />).find('RatingInput').length).toEqual(5);
  });

  it('should render five <RatingLabel/>', () => {
    expect(shallow(<Rating />).find('RatingLabel').length).toEqual(5);
  });

  it('should render <RatingTotal/> if readonly is true', () => {
    expect(shallow(<Rating readonly />).find('RatingTotal').length).toEqual(1);
  });

  it('should render <SubmitRating/> if no readonly is undefined', () => {
    expect(shallow(<Rating />).find('SubmitRating').length).toEqual(1);
  });

  it('should not render <SubmitRating/> if readonly is true', () => {
    expect(shallow(<Rating readonly />).find('SubmitRating').length).toEqual(0);
  });

  it('should default to no ratings', () => {
    const wrapper = shallow(<Rating />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'totalRatings');
    expect(instance.totalRatings()).toEqual('No ratings yet.');

    // TODO: Test the component functions
  });
});
