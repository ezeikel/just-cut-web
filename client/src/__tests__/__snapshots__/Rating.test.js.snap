// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating Component should match snapshot 1`] = `
<RatingWrapper
  className=""
>
  <RatingInput
    id="1"
    type="radio"
    value="1"
  />
  <RatingLabel
    className=""
    htmlFor="1"
  >
    Abysmal
  </RatingLabel>
  <RatingInput
    id="2"
    type="radio"
    value="2"
  />
  <RatingLabel
    className=""
    htmlFor="2"
  >
    Poor
  </RatingLabel>
  <RatingInput
    id="3"
    type="radio"
    value="3"
  />
  <RatingLabel
    className=""
    htmlFor="3"
  >
    Alright
  </RatingLabel>
  <RatingInput
    id="4"
    type="radio"
    value="4"
  />
  <RatingLabel
    className=""
    htmlFor="4"
  >
    Good
  </RatingLabel>
  <RatingInput
    id="5"
    type="radio"
    value="5"
  />
  <RatingLabel
    className=""
    htmlFor="5"
  >
    Great
  </RatingLabel>
  <SubmitRating
    className=""
  >
    Submit Rating
  </SubmitRating>
</RatingWrapper>
`;
